dss/api/search.py

```diff
@@ -1,42 +1,142 @@
-        search_obj = Search(using=es_client,
-                            index=Config.get_es_index_name(ESIndexType.docs, Replica[replica]),
-                            doc_type=ESDocType.doc.name).update_from_dict(es_query)
+        if _scroll_id is None:
+            page = es_client.search(index=Config.get_es_index_name(ESIndexType.docs, Replica[replica]),
+                                    doc_type=ESDocType.doc.name,
+                                    scroll=scroll,
+                                    size=per_page,
+                                    body=es_query,
+                                    sort=sort
+                                    )
+            get_logger().debug("Created ES scroll instance")
+        else:
+            get_logger().debug("Retrieve ES results from scroll instance Scroll_id: %s", _scroll_id)
+            page = es_client.scroll(scroll_id=_scroll_id, scroll=scroll)
 
-        # TODO (mbaumann) extract version from the request path instead of hard-coding it here
-        bundles_url_base = request.host_url + 'v1/bundles/'
+        # TODO: (tsmith12) allow users to retrieve previous search results
+        _scroll_id = page['_scroll_id']
         result_list = [{
-            'bundle_id': hit.meta.id,
-            # TODO:  use dss.util.UrlBuilder for bundle_url
-            'bundle_url': bundles_url_base + hit.meta.id.replace(".", "?version=", 1) + f"&replica={replica}",
-            'search_score': hit.meta.score
-        } for hit in search_obj.scan()]
-        return jsonify({'es_query': es_query, 'results': result_list})
-
-    except ElasticsearchDslException:
-        raise DSSException(requests.codes.bad_request,
-                           "elasticsearch_query_error",
-                           "Invalid query")
-    except ElasticsearchException:
+            'bundle_id': hit['_id'],
+            'bundle_url': _build_bundle_url(hit, replica),
+            'search_score': hit['_score']
+        } for hit in page['hits']['hits']]
+
+        # TODO: (tsmith12) if page returns 0 hits, then all results have been found. delete search id
+        # TODO: (tsmith12) if all results found return request.code.ok.
+        if len(result_list) < per_page:
+            links = ''
+        else:
+            next_url = request.host_url + str(UrlBuilder().set(path="v1/search")
+                                              .add_query('per_page', str(per_page))
+                                              .add_query("replica", replica)
+                                              .add_query("_scroll_id", _scroll_id))
+            links = build_link_header({next_url: {"rel": "next"}})
+
+        # TODO: (tsmith12) check if all results found and return request.code.ok.
+        # TODO: (tsmith12) if all results not found return request.code.partial.
+        response = make_response(jsonify({'es_query': es_query, 'results': result_list}), requests.codes.ok)
+        response.headers['Link'] = links
```

Comment:
If `links` is empty, then there should be no `Link` header. Or does that not pass validation?

Reply:
It wont pass validation. I could return `Partial 206` when there are more result to retrieve and that can have Link in the header. Then I could return `Ok 200` when all results have been returned?

Reply:
I add 206 response to search on this branch https://github.com/HumanCellAtlas/data-store/tree/tsmith-partial-search